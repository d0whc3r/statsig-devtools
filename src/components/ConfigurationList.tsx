import React from 'react'

import {
  ConfigurationEmptyState,
  ConfigurationErrorState,
  ConfigurationItem,
  ConfigurationLoadingState,
  ConfigurationSearchAndFilters,
  useConfigurationFilters,
  useConfigurationOverrides,
} from './configuration-list'

import type { EvaluationResult } from '../services/unified-statsig-api'
import type { StatsigConfigurationItem, StorageOverride } from '../types'

interface ConfigurationListProps {
  configurations: StatsigConfigurationItem[]
  evaluationResults: Map<string, EvaluationResult>
  activeOverrides?: StorageOverride[]
  onConfigurationSelect: (config: StatsigConfigurationItem) => void
  selectedConfiguration?: StatsigConfigurationItem
  isLoading?: boolean
  error?: string
  viewMode?: 'popup' | 'sidebar' | 'tab'
}

/**
 * Configuration list component with search, filter, and virtual scrolling
 */
export const ConfigurationList = React.memo(
  ({
    configurations,
    evaluationResults,
    activeOverrides = [],
    onConfigurationSelect,
    selectedConfiguration,
    isLoading = false,
    error,
    viewMode: _viewMode = 'popup',
  }: ConfigurationListProps) => {
    // Use custom hooks for filtering and overrides logic
    const {
      searchQuery,
      filterType,
      filterStatus,
      filteredConfigurations,
      handleSearchChange,
      setFilterType,
      setFilterStatus,
      clearFilters,
      hasActiveFilters,
    } = useConfigurationFilters(configurations, evaluationResults)

    const { hasOverrides, getOverrideCount } = useConfigurationOverrides(activeOverrides)

    // Handle error state
    if (error) {
      return <ConfigurationErrorState error={error} />
    }

    return (
      <div className="flex h-full flex-col">
        {/* Search and Filter Controls */}
        <div className="relative z-30">
          <ConfigurationSearchAndFilters
            searchQuery={searchQuery}
            filterType={filterType}
            filterStatus={filterStatus}
            onSearchChange={handleSearchChange}
            onFilterTypeChange={setFilterType}
            onFilterStatusChange={setFilterStatus}
            onClearFilters={clearFilters}
            hasActiveFilters={hasActiveFilters}
            totalConfigurations={configurations.length}
            filteredCount={filteredConfigurations.length}
          />
        </div>

        {/* Configuration List */}
        <div
          className={`custom-scrollbar min-h-0 flex-1 ${_viewMode === 'sidebar' ? 'overflow-visible' : 'overflow-y-auto'}`}
        >
          {isLoading ? (
            <ConfigurationLoadingState />
          ) : filteredConfigurations.length === 0 ? (
            <ConfigurationEmptyState hasActiveFilters={hasActiveFilters} />
          ) : (
            <div className="py-2">
              {filteredConfigurations.map((config) => {
                const isSelected = selectedConfiguration?.name === config.name
                const result = evaluationResults.get(config.name)
                const configHasOverrides = hasOverrides(config.name)
                const overrideCount = getOverrideCount(config.name)

                return (
                  <ConfigurationItem
                    key={config.name}
                    config={config}
                    isSelected={isSelected}
                    result={result}
                    onSelect={onConfigurationSelect}
                    hasOverrides={configHasOverrides}
                    overrideCount={overrideCount}
                  />
                )
              })}
            </div>
          )}
        </div>
      </div>
    )
  },
)

ConfigurationList.displayName = 'ConfigurationList'
