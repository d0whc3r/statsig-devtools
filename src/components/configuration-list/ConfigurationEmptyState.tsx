import React from 'react'

interface ConfigurationEmptyStateProps {
  hasActiveFilters: boolean
}

/**
 * Empty state component for configuration list
 */
export function ConfigurationEmptyState({ hasActiveFilters }: ConfigurationEmptyStateProps) {
  return (
    <div className="p-8 text-center">
      <div className="text-gray-400">
        <svg className="mx-auto h-12 w-12" fill="none" viewBox="0 0 24 24" stroke="currentColor">
          <path
            strokeLinecap="round"
            strokeLinejoin="round"
            strokeWidth={2}
            d="M9 5H7a2 2 0 00-2 2v10a2 2 0 002 2h8a2 2 0 002-2V7a2 2 0 00-2-2h-2M9 5a2 2 0 002 2h2a2 2 0 002-2M9 5a2 2 0 012-2h2a2 2 0 012 2"
          />
        </svg>
      </div>
      <h3 className="mt-2 text-sm font-medium text-gray-900">No configurations found</h3>
      <p className="mt-1 text-sm text-gray-500">
        {hasActiveFilters ? 'Try adjusting your search or filters.' : 'No configurations available.'}
      </p>
    </div>
  )
}
