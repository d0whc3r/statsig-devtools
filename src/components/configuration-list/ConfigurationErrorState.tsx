import React from 'react'

interface ConfigurationErrorStateProps {
  error: string
}

/**
 * Error state component for configuration list
 */
export function ConfigurationErrorState({ error }: ConfigurationErrorStateProps) {
  return (
    <div className="p-4">
      <div className="rounded-md bg-red-50 p-4">
        <div className="flex">
          <div className="ml-3">
            <h3 className="text-sm font-medium text-red-800">Error loading configurations</h3>
            <div className="mt-2 text-sm text-red-700">
              <p>{error}</p>
            </div>
          </div>
        </div>
      </div>
    </div>
  )
}
