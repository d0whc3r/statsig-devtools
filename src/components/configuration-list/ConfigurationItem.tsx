import React, { useCallback } from 'react'

import { formatTypeName, formatValue, getTypeBadgeClass } from '../../utils/configuration-formatters'
import { ConfigurationStatusIndicator } from '../ConfigurationStatusIndicator'

import type { EvaluationResult } from '../../services/statsig-integration'
import type { StatsigConfigurationItem } from '../../types'

interface ConfigurationItemProps {
  config: StatsigConfigurationItem
  isSelected: boolean
  result?: EvaluationResult
  onSelect: (config: StatsigConfigurationItem) => void
  hasOverrides: boolean
  overrideCount: number
}

/**
 * Individual configuration item component
 */
export const ConfigurationItem = React.memo(
  ({ config, isSelected, result, onSelect, hasOverrides, overrideCount }: ConfigurationItemProps) => {
    const handleClick = useCallback(() => {
      onSelect(config)
    }, [config, onSelect])

    const handleKeyDown = useCallback(
      (event: React.KeyboardEvent) => {
        if (event.key === 'Enter' || event.key === ' ') {
          event.preventDefault()
          handleClick()
        }
      },
      [handleClick],
    )

    return (
      <div
        onClick={handleClick}
        onKeyDown={handleKeyDown}
        role="button"
        tabIndex={0}
        className={`card-professional mx-2 my-1.5 cursor-pointer p-3 ${isSelected ? 'selected' : ''}`}
      >
        <div className="flex items-start justify-between">
          <div className="min-w-0 flex-1">
            <div className="mb-2 flex items-center gap-2">
              <ConfigurationStatusIndicator result={result} size="sm" />
              <h3 className="truncate text-sm font-semibold text-gray-900">{config.name}</h3>
            </div>

            <div className="mb-2 flex flex-wrap items-center gap-1">
              <span className={`${getTypeBadgeClass(config.type)} text-xs`}>{formatTypeName(config.type)}</span>

              {!config.enabled && (
                <span className="type-badge bg-gray-100 text-xs text-gray-800">
                  <svg className="h-2.5 w-2.5" fill="none" viewBox="0 0 24 24" stroke="currentColor">
                    <path
                      strokeLinecap="round"
                      strokeLinejoin="round"
                      strokeWidth={2}
                      d="M18.364 18.364A9 9 0 005.636 5.636m12.728 12.728L5.636 5.636m12.728 12.728L18.364 5.636M5.636 18.364l12.728-12.728"
                    />
                  </svg>
                  Disabled
                </span>
              )}

              {hasOverrides && (
                <span className="type-badge border border-orange-200 bg-orange-100 text-xs text-orange-800">
                  <svg className="h-2.5 w-2.5" fill="none" viewBox="0 0 24 24" stroke="currentColor">
                    <path
                      strokeLinecap="round"
                      strokeLinejoin="round"
                      strokeWidth={2}
                      d="M12 9v2m0 4h.01m-6.938 4h13.856c1.54 0 2.502-1.667 1.732-2.5L13.732 4c-.77-.833-1.964-.833-2.732 0L3.732 16.5c-.77.833.192 2.5 1.732 2.5z"
                    />
                  </svg>
                  {overrideCount}
                </span>
              )}
            </div>

            {result && (
              <div className="rounded-md bg-gray-50 px-2 py-1.5 text-xs text-gray-600">
                {config.type === 'feature_gate' && (
                  <span className="font-medium">
                    <span className={result.value ? 'text-green-700' : 'text-red-700'}>
                      {formatValue(result.value)}
                    </span>
                  </span>
                )}
                {config.type === 'experiment' && result.groupName && (
                  <span className="font-medium">
                    Group: <span className="text-purple-700">{result.groupName}</span>
                  </span>
                )}
                {config.type === 'dynamic_config' && <span className="font-medium text-green-700">Ready</span>}
              </div>
            )}
          </div>

          <div className="ml-2 flex-shrink-0">
            <svg className="h-4 w-4 text-gray-400" fill="none" viewBox="0 0 24 24" stroke="currentColor">
              <path strokeLinecap="round" strokeLinejoin="round" strokeWidth={2} d="M9 5l7 7-7 7" />
            </svg>
          </div>
        </div>
      </div>
    )
  },
)

ConfigurationItem.displayName = 'ConfigurationItem'
