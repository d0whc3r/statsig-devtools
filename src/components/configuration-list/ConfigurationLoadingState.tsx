import React from 'react'

import { LoadingSpinner } from '../LoadingSpinner'

/**
 * Loading state component for configuration list
 */
export function ConfigurationLoadingState() {
  return (
    <div className="flex items-center justify-center p-8">
      <div className="text-center">
        <LoadingSpinner size="md" />
        <p className="mt-2 text-sm text-gray-600">Loading configurations...</p>
      </div>
    </div>
  )
}
