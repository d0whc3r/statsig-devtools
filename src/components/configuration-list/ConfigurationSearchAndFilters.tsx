import React from 'react'

import type { FilterStatus, FilterType } from './useConfigurationFilters'

interface ConfigurationSearchAndFiltersProps {
  searchQuery: string
  filterType: FilterType
  filterStatus: FilterStatus
  onSearchChange: (event: React.ChangeEvent<HTMLInputElement>) => void
  onFilterTypeChange: (filterType: FilterType) => void
  onFilterStatusChange: (filterStatus: FilterStatus) => void
  onClearFilters: () => void
  hasActiveFilters: boolean
  totalConfigurations: number
  filteredCount: number
}

/**
 * Search and filter controls component
 */
export function ConfigurationSearchAndFilters({
  searchQuery,
  filterType,
  filterStatus,
  onSearchChange,
  onFilterTypeChange,
  onFilterStatusChange,
  onClearFilters,
  hasActiveFilters,
  totalConfigurations,
  filteredCount,
}: ConfigurationSearchAndFiltersProps) {
  return (
    <div className="flex-shrink-0 border-b border-gray-200 p-3">
      <div className="mb-3">
        <div className="relative">
          <div className="pointer-events-none absolute inset-y-0 left-0 flex items-center pl-3">
            <svg className="h-4 w-4 text-slate-400" fill="none" viewBox="0 0 24 24" stroke="currentColor">
              <path
                strokeLinecap="round"
                strokeLinejoin="round"
                strokeWidth={2}
                d="M21 21l-6-6m2-5a7 7 0 11-14 0 7 7 0 0114 0z"
              />
            </svg>
          </div>
          <input
            type="text"
            placeholder="Search..."
            value={searchQuery}
            onChange={onSearchChange}
            className="input-professional w-full py-2.5 pr-4 pl-10 text-sm"
          />
        </div>
      </div>

      <div className="flex items-center gap-2">
        <select
          value={filterType}
          onChange={(e) => onFilterTypeChange(e.target.value as FilterType)}
          className="input-professional min-w-0 flex-1 px-3 py-1.5 text-xs font-medium"
        >
          <option value="all">All Types</option>
          <option value="feature_gate">üö™ Gates</option>
          <option value="experiment">üß™ Experiments</option>
          <option value="dynamic_config">‚öôÔ∏è Configs</option>
        </select>

        <select
          value={filterStatus}
          onChange={(e) => onFilterStatusChange(e.target.value as FilterStatus)}
          className="input-professional min-w-0 flex-1 px-3 py-1.5 text-xs font-medium"
        >
          <option value="all">All Status</option>
          <option value="passed">‚úÖ Passed</option>
          <option value="failed">‚ùå Failed</option>
        </select>

        {hasActiveFilters && (
          <button
            onClick={onClearFilters}
            className="icon-button-professional h-8 w-8 flex-shrink-0 text-slate-600 hover:text-red-600"
            title="Clear filters"
          >
            <svg className="h-3.5 w-3.5" fill="none" viewBox="0 0 24 24" stroke="currentColor">
              <path strokeLinecap="round" strokeLinejoin="round" strokeWidth={2} d="M6 18L18 6M6 6l12 12" />
            </svg>
          </button>
        )}
      </div>

      <div className="mt-3 flex items-center justify-between">
        <div className="flex items-center gap-2 text-xs text-slate-600">
          <svg className="h-3.5 w-3.5" fill="none" viewBox="0 0 24 24" stroke="currentColor">
            <path
              strokeLinecap="round"
              strokeLinejoin="round"
              strokeWidth={2}
              d="M9 5H7a2 2 0 00-2 2v10a2 2 0 002 2h8a2 2 0 002-2V7a2 2 0 00-2-2h-2M9 5a2 2 0 002 2h2a2 2 0 002-2M9 5a2 2 0 012-2h2a2 2 0 012 2"
            />
          </svg>
          <span className="font-medium">
            {filteredCount === totalConfigurations ? (
              <span>{totalConfigurations} configurations</span>
            ) : (
              <span>
                <span className="font-semibold text-blue-600">{filteredCount}</span> of {totalConfigurations}
              </span>
            )}
          </span>
        </div>
        {hasActiveFilters && <div className="text-xs font-medium text-orange-600">Filtered</div>}
      </div>
    </div>
  )
}
