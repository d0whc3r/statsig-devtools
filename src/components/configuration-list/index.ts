export { ConfigurationEmptyState } from './ConfigurationEmptyState'
export { ConfigurationErrorState } from './ConfigurationErrorState'
export { ConfigurationItem } from './ConfigurationItem'
export { ConfigurationLoadingState } from './ConfigurationLoadingState'
export { ConfigurationSearchAndFilters } from './ConfigurationSearchAndFilters'
export type { FilterStatus, FilterType } from './useConfigurationFilters'
export { useConfigurationFilters } from './useConfigurationFilters'
export { useConfigurationOverrides } from './useConfigurationOverrides'
