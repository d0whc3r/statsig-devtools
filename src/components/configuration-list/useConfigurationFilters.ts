import { useCallback, useMemo, useState } from 'react'

import { useDebouncedSearch } from '../VirtualizedList'

import type { EvaluationResult } from '../../services/unified-statsig-api'
import type { StatsigConfigurationItem } from '../../types'
import type React from 'react'

export type FilterType = 'all' | 'feature_gate' | 'experiment' | 'dynamic_config'
export type FilterStatus = 'all' | 'passed' | 'failed'

/**
 * Hook to manage configuration filtering and searching
 */
export const useConfigurationFilters = (
  configurations: StatsigConfigurationItem[],
  evaluationResults: Map<string, EvaluationResult>,
) => {
  const [debouncedSearchQuery, setSearchQuery, immediateSearchQuery] = useDebouncedSearch('', 300)
  const [filterType, setFilterType] = useState<FilterType>('all')
  const [filterStatus, setFilterStatus] = useState<FilterStatus>('all')

  /**
   * Filter configurations by type
   */
  const filterByType = useCallback(
    (configs: StatsigConfigurationItem[]) =>
      filterType === 'all' ? configs : configs.filter((config) => config.type === filterType),
    [filterType],
  )

  /**
   * Filter configurations by evaluation status
   */
  const filterByStatus = useCallback(
    (configs: StatsigConfigurationItem[]) => {
      if (filterStatus === 'all') return configs

      return configs.filter((config) => {
        const result = evaluationResults.get(config.name)
        if (!result) return filterStatus === 'failed'
        return filterStatus === 'passed' ? result.passed : !result.passed
      })
    },
    [filterStatus, evaluationResults],
  )

  /**
   * Filter configurations by search query (searches both name and ID)
   */
  const filterBySearch = useCallback(
    (configs: StatsigConfigurationItem[]) => {
      if (!debouncedSearchQuery.trim()) return configs

      const query = debouncedSearchQuery.toLowerCase().trim()
      return configs.filter((config) => {
        const nameMatch = config.name.toLowerCase().includes(query)
        const idMatch = config.id ? config.id.toLowerCase().includes(query) : false
        return nameMatch || idMatch
      })
    },
    [debouncedSearchQuery],
  )

  /**
   * Filter and search configurations
   */
  const filteredConfigurations = useMemo(() => {
    let filtered = configurations
    filtered = filterByType(filtered)
    filtered = filterByStatus(filtered)
    filtered = filterBySearch(filtered)
    return filtered
  }, [configurations, filterByType, filterByStatus, filterBySearch])

  /**
   * Handle search input change with debouncing
   */
  const handleSearchChange = useCallback(
    (event: React.ChangeEvent<HTMLInputElement>) => {
      setSearchQuery(event.target.value)
    },
    [setSearchQuery],
  )

  /**
   * Clear all filters
   */
  const clearFilters = useCallback(() => {
    setSearchQuery('')
    setFilterType('all')
    setFilterStatus('all')
  }, [setSearchQuery])

  /**
   * Check if any filters are active
   */
  const hasActiveFilters = useMemo(
    () => immediateSearchQuery.trim() !== '' || filterType !== 'all' || filterStatus !== 'all',
    [immediateSearchQuery, filterType, filterStatus],
  )

  return {
    searchQuery: immediateSearchQuery,
    filterType,
    filterStatus,
    filteredConfigurations,
    handleSearchChange,
    setFilterType,
    setFilterStatus,
    clearFilters,
    hasActiveFilters,
  }
}
