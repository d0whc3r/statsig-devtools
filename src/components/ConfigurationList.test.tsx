import { beforeEach, describe, expect, it, vi } from 'vitest'

import * as configListModule from './configuration-list'
import { ConfigurationList } from './ConfigurationList'

import type { EvaluationResult } from '../services/unified-statsig-api'
import type { StatsigConfigurationItem, StorageOverride } from '../types'

import { render, screen } from '@testing-library/react'

// Mock the configuration-list components
vi.mock('./configuration-list', () => ({
  ConfigurationEmptyState: ({ hasActiveFilters }: { hasActiveFilters: boolean }) => (
    <div data-testid="empty-state">{hasActiveFilters ? 'No configurations match filters' : 'No configurations'}</div>
  ),
  ConfigurationErrorState: ({ error }: { error: string }) => <div data-testid="error-state">{error}</div>,
  ConfigurationItem: ({ config, isSelected, onSelect }: any) => (
    <button
      data-testid={`config-item-${config.name}`}
      onClick={() => onSelect(config)}
      className={isSelected ? 'selected' : ''}
    >
      {config.name}
    </button>
  ),
  ConfigurationLoadingState: () => <div data-testid="loading-state">Loading...</div>,
  ConfigurationSearchAndFilters: ({ searchQuery, totalConfigurations, filteredCount }: any) => (
    <div data-testid="search-filters">
      Search: {searchQuery}, Total: {totalConfigurations}, Filtered: {filteredCount}
    </div>
  ),
  useConfigurationFilters: vi.fn(),
  useConfigurationOverrides: vi.fn(),
}))

const mockUseConfigurationFilters = vi.mocked(configListModule.useConfigurationFilters)
const mockUseConfigurationOverrides = vi.mocked(configListModule.useConfigurationOverrides)

describe('ConfigurationList', () => {
  const mockConfigurations: StatsigConfigurationItem[] = [
    {
      name: 'test_gate',
      type: 'feature_gate',
      enabled: true,
    },
    {
      name: 'test_config',
      type: 'dynamic_config',
      enabled: true,
    },
  ]

  const mockEvaluationResults = new Map<string, EvaluationResult>([
    [
      'test_gate',
      {
        configurationName: 'test_gate',
        type: 'feature_gate',
        passed: true,
        value: true,
        ruleId: 'rule1',
        groupName: 'group1',
        reason: 'Test reason',
      },
    ],
    [
      'test_config',
      {
        configurationName: 'test_config',
        type: 'dynamic_config',
        passed: true,
        value: { key: 'value' },
        ruleId: 'rule2',
        groupName: 'group2',
        reason: 'Test reason',
      },
    ],
  ])

  const mockActiveOverrides: StorageOverride[] = [
    {
      type: 'localStorage',
      key: 'test_gate',
      value: 'false',
      featureName: 'test_gate',
      featureType: 'feature_gate',
    },
  ]

  const mockOnConfigurationSelect = vi.fn()

  beforeEach(() => {
    // Default mock implementations
    mockUseConfigurationFilters.mockReturnValue({
      searchQuery: '',
      filterType: 'all',
      filterStatus: 'all',
      filteredConfigurations: mockConfigurations,
      handleSearchChange: vi.fn(),
      setFilterType: vi.fn(),
      setFilterStatus: vi.fn(),
      clearFilters: vi.fn(),
      hasActiveFilters: false,
    })

    mockUseConfigurationOverrides.mockReturnValue({
      hasOverrides: vi.fn().mockReturnValue(false),
      getOverrideCount: vi.fn().mockReturnValue(0),
    })
  })

  it('should match snapshot with default state', () => {
    const { container } = render(
      <ConfigurationList
        configurations={mockConfigurations}
        evaluationResults={mockEvaluationResults}
        onConfigurationSelect={mockOnConfigurationSelect}
      />,
    )
    expect(container).toMatchSnapshot()
  })

  it('should match snapshot with loading state', () => {
    const { container } = render(
      <ConfigurationList
        configurations={mockConfigurations}
        evaluationResults={mockEvaluationResults}
        onConfigurationSelect={mockOnConfigurationSelect}
        isLoading
      />,
    )
    expect(container).toMatchSnapshot()
  })

  it('should match snapshot with error state', () => {
    const { container } = render(
      <ConfigurationList
        configurations={mockConfigurations}
        evaluationResults={mockEvaluationResults}
        onConfigurationSelect={mockOnConfigurationSelect}
        error="Failed to load configurations"
      />,
    )
    expect(container).toMatchSnapshot()
  })

  it('renders loading state when isLoading is true', () => {
    render(
      <ConfigurationList
        configurations={mockConfigurations}
        evaluationResults={mockEvaluationResults}
        onConfigurationSelect={mockOnConfigurationSelect}
        isLoading
      />,
    )

    expect(screen.getByTestId('loading-state')).toBeInTheDocument()
    expect(screen.getByText('Loading...')).toBeInTheDocument()
  })

  it('renders error state when error is provided', () => {
    const errorMessage = 'Failed to load configurations'

    render(
      <ConfigurationList
        configurations={mockConfigurations}
        evaluationResults={mockEvaluationResults}
        onConfigurationSelect={mockOnConfigurationSelect}
        error={errorMessage}
      />,
    )

    expect(screen.getByTestId('error-state')).toBeInTheDocument()
    expect(screen.getByText(errorMessage)).toBeInTheDocument()
  })

  it('renders empty state when no configurations match filters', () => {
    mockUseConfigurationFilters.mockReturnValue({
      searchQuery: 'nonexistent',
      filterType: 'all',
      filterStatus: 'all',
      filteredConfigurations: [],
      handleSearchChange: vi.fn(),
      setFilterType: vi.fn(),
      setFilterStatus: vi.fn(),
      clearFilters: vi.fn(),
      hasActiveFilters: true,
    })

    render(
      <ConfigurationList
        configurations={mockConfigurations}
        evaluationResults={mockEvaluationResults}
        onConfigurationSelect={mockOnConfigurationSelect}
      />,
    )

    expect(screen.getByTestId('empty-state')).toBeInTheDocument()
    expect(screen.getByText('No configurations match filters')).toBeInTheDocument()
  })

  it('renders configuration items when configurations are available', () => {
    render(
      <ConfigurationList
        configurations={mockConfigurations}
        evaluationResults={mockEvaluationResults}
        onConfigurationSelect={mockOnConfigurationSelect}
      />,
    )

    expect(screen.getByTestId('config-item-test_gate')).toBeInTheDocument()
    expect(screen.getByTestId('config-item-test_config')).toBeInTheDocument()
    expect(screen.getByText('test_gate')).toBeInTheDocument()
    expect(screen.getByText('test_config')).toBeInTheDocument()
  })

  it('renders search and filters component', () => {
    render(
      <ConfigurationList
        configurations={mockConfigurations}
        evaluationResults={mockEvaluationResults}
        onConfigurationSelect={mockOnConfigurationSelect}
      />,
    )

    expect(screen.getByTestId('search-filters')).toBeInTheDocument()
    expect(screen.getByText(/Total: 2, Filtered: 2/)).toBeInTheDocument()
  })

  it('handles configuration selection', () => {
    render(
      <ConfigurationList
        configurations={mockConfigurations}
        evaluationResults={mockEvaluationResults}
        onConfigurationSelect={mockOnConfigurationSelect}
      />,
    )

    const configItem = screen.getByTestId('config-item-test_gate')
    configItem.click()

    expect(mockOnConfigurationSelect).toHaveBeenCalledWith(mockConfigurations[0])
  })

  it('shows selected configuration with correct styling', () => {
    render(
      <ConfigurationList
        configurations={mockConfigurations}
        evaluationResults={mockEvaluationResults}
        onConfigurationSelect={mockOnConfigurationSelect}
        selectedConfiguration={mockConfigurations[0]}
      />,
    )

    const selectedItem = screen.getByTestId('config-item-test_gate')
    expect(selectedItem).toHaveClass('selected')
  })

  it('handles active overrides correctly', () => {
    mockUseConfigurationOverrides.mockReturnValue({
      hasOverrides: vi.fn().mockImplementation((name) => name === 'test_gate'),
      getOverrideCount: vi.fn().mockImplementation((name) => (name === 'test_gate' ? 1 : 0)),
    })

    render(
      <ConfigurationList
        configurations={mockConfigurations}
        evaluationResults={mockEvaluationResults}
        activeOverrides={mockActiveOverrides}
        onConfigurationSelect={mockOnConfigurationSelect}
      />,
    )

    // The component should render configuration items
    expect(screen.getAllByTestId('config-item-test_gate')).toHaveLength(1)
    expect(screen.getAllByTestId('config-item-test_config')).toHaveLength(1)
  })
})
