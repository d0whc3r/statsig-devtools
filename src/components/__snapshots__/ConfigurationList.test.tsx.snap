// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ConfigurationList > should match snapshot with default state 1`] = `
<div>
  <div
    class="flex h-full flex-col"
  >
    <div
      class="relative z-30"
    >
      <div
        data-testid="search-filters"
      >
        Search: 
        , Total: 
        2
        , Filtered: 
        2
      </div>
    </div>
    <div
      class="custom-scrollbar min-h-0 flex-1 overflow-y-auto"
    >
      <div
        class="py-2"
      >
        <button
          class=""
          data-testid="config-item-test_gate"
        >
          test_gate
        </button>
        <button
          class=""
          data-testid="config-item-test_config"
        >
          test_config
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`ConfigurationList > should match snapshot with error state 1`] = `
<div>
  <div
    data-testid="error-state"
  >
    Failed to load configurations
  </div>
</div>
`;

exports[`ConfigurationList > should match snapshot with loading state 1`] = `
<div>
  <div
    class="flex h-full flex-col"
  >
    <div
      class="relative z-30"
    >
      <div
        data-testid="search-filters"
      >
        Search: 
        , Total: 
        2
        , Filtered: 
        2
      </div>
    </div>
    <div
      class="custom-scrollbar min-h-0 flex-1 overflow-y-auto"
    >
      <div
        data-testid="loading-state"
      >
        Loading...
      </div>
    </div>
  </div>
</div>
`;
